### PR TITLE
Add core classes and functions for online clustering diarizer part 1

### PR DESCRIPTION
**Pull request by tango4j**
Signed-off-by: Taejin Park <tango4j@gmail.com>

[[NeMo Online Speaker Diarization VIDEO DEMO LINK](
https://drive.google.com/file/d/15ro8HKmzevXh_wx9oTtfRUw9wNtj6LuJ/view?usp=share_link)]

# What does this PR do ?

Add core classes and functions for online clustering diarizer.
Dividing the huge PR into two: 

part 1: online clustering (nmesc_clustering.py) + unit tests
part 2: Segmentation + online diarizer class + unit tests

**Collection**: 

ASR

# Changelog 

#### `OnlineSpeakerClustering` in nmesc_clustering.py (included in this PR)
online version of `SpeakerClustering` class (online speaker counting and permutation matching)

#### `test_diar_utils.py` in NeMo/tests/collections/asr/  (included in this PR)
Tests all clustering and online diarization related functions

-------------
the following will be included in the later PR
#### `OnlineClusteringDiarizer` in online_diarizer.py (inherits ClusteringDiarizer) (will be included in part2)
online version of `ClusteringDiarizer` class in clustering_diarizer.py

#### `OnlineSegmentor` class in speaker_utils.py: (will be included in part2)
online version of speech segmentations

image

Checkout [Help Slides Link](https://docs.google.com/presentation/d/1LDUTIGRNVuLJsy81V3GaHZlFDUpxZ9rux-fNPfgcLmw/edit?usp=sharing) for better understanding on why online diarizer needs
(1) linear sum assignment (Hungarian alg)
(2) Interval merging
(3) merge_vector function
and data flow of class and functions.

image

If you are curious about why online diarization needs FIFO-queue, take a look at the most recent paper on online neural diarization
[Online Neural Diarization of Unlimited Numbers of Speakers](https://arxiv.org/pdf/2206.02432.pdf)

# Usage

This is core classes and functions update.
Example script will be pushed in the second round of PR.

# Before your PR is "Ready for review"
**Pre checks**:
- [x] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [x] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:
- [x] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.
[Online Neural Diarization of Unlimited Numbers of Speakers](https://arxiv.org/pdf/2206.02432.pdf)

## Who can review?

Anyone in the community is free to review the PR once the checks have passed. 
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information
* Related to # (issue)

### CODE REVIEW

File: tests/collections/asr/test_diar_utils.py [hunk 1]
```diff
@@ -35,130 +45,289 @@ def check_range_values(target, source):
...
     torch_seed=0,
 ):
-
     torch.manual_seed(torch_seed)
-    segment_lists = []
     spk_timestamps = [(spk_dur * k, spk_dur) for k in range(n_spks)]
     emb_list, seg_list = [], []
     multiscale_segment_counts = [0 for _ in range(len(ms_window))]
+    ground_truth = []
+    random_orhtogonal_embs = generate_orthogonal_embs(n_spks, perturb_sigma, emb_dim)
```
Comment (SeanNaren) [inline tests/collections/asr/test_diar_utils.py:89]: Spelling mistake, annoying word :D 
```suggestion
    random_orthogonal_embs = generate_orthogonal_embs(n_spks, perturb_sigma, emb_dim)
```

File: nemo/collections/asr/parts/utils/nmesc_clustering.py [hunk 2]
```diff
@@ -1253,3 +1787,612 @@ def forward_infer(
         spectral_model = SpectralClustering(n_clusters=n_clusters, cuda=self.cuda, device=self.device)
         Y = spectral_model.forward(affinity_mat)
         return Y
+
+
+class OnlineSpeakerClustering:
```
Comment (SeanNaren) [inline nemo/collections/asr/parts/utils/nmesc_clustering.py:1792]: This file (`utils/nmesc_clustering.py`) is massive, should we create a folder within this general NeMo directory and split things out? i.e

`nemo/collections/asr/parts/utils/diarization/online_clustering.py`

File: nemo/collections/asr/parts/utils/nmesc_clustering.py [hunk 3]
```diff
@@ -1253,3 +1787,612 @@ def forward_infer(
...
+                    history buffer.
+            history_embedding_buffer_emb (Tensor)
+                Tensor containing speaker embedding vectors for saving the history of the previous
+                speaker profile in the given audio session
+            history_embedding_buffer_label (Tensor)
+                Speaker label (cluster label) for embedding vectors saved in the history buffer
+            Y_fullhist (Tensor)
+                Tensor containing the speaker label hypothesis from start to current frame
+        """
+        self.isOnline = False
```
Comment (SeanNaren) [inline nemo/collections/asr/parts/utils/nmesc_clustering.py:1943]:
```suggestion
        self.is_online = False
```

File: tests/collections/asr/test_diar_utils.py [hunk 4]
```diff
@@ -35,130 +45,289 @@ def check_range_values(target, source):
...
+    return orth_embs
 
 
-def generate_mock_data(
+def generate_toy_data(
     n_spks=2,
     spk_dur=3,
     emb_dim=192,
-    perturb_sigma=0.01,
+    perturb_sigma=0.0,
```
[the remark this answers is not in the document]
Comment (tango4j) [reply tests/collections/asr/test_diar_utils.py:79]: I noticed that very rarely even perturb_sigma=0.01 can cause failure of speaker counting, so I switched this to 0.0 as a default value when I use toy data for sanity checking.

File: tests/collections/asr/test_diar_utils.py (hunk 1 above)
Comment (tango4j) [reply tests/collections/asr/test_diar_utils.py:89]: fixed

File: nemo/collections/asr/parts/utils/nmesc_clustering.py (hunk 2 above)
Comment (tango4j) [reply nemo/collections/asr/parts/utils/nmesc_clustering.py:1792]: changed nmesc_clustering.py to offline_clustering.py then added online_clustering.py for all the newly added online clustering modules

File: nemo/collections/asr/parts/utils/nmesc_clustering.py (hunk 3 above)
Comment (tango4j) [reply nemo/collections/asr/parts/utils/nmesc_clustering.py:1943]: changed

File: nemo/collections/asr/parts/utils/online_clustering.py [hunk 5]
```diff
@@ -0,0 +1,1931 @@
...
+            n_clusters = int(est_num_of_spk_enhanced.item())
+        else:
+            n_clusters = int(est_num_of_spk.item())
+
+        spectral_model = SpectralClustering(n_clusters=n_clusters, cuda=self.cuda, device=self.device)
+        Y = spectral_model.forward(affinity_mat)
+        return Y
+
+
+class OnlineSpeakerClustering:
```
[the remark this answers is not in the document]
Comment (tango4j) [reply nemo/collections/asr/parts/utils/online_clustering.py:716]: resolved

File: nemo/collections/asr/parts/utils/offline_clustering.py [hunk 6]
```diff
@@ -33,12 +33,16 @@
 
 from typing import Dict, List, Tuple
 
+import numpy as np
```
[the remark this answers is not in the document]
Comment (tango4j) [reply nemo/collections/asr/parts/utils/offline_clustering.py:36]: fixed

File: nemo/collections/asr/parts/utils/online_clustering.py [hunk 7]
```diff
@@ -0,0 +1,1106 @@
...
+    index_mapping: Tuple[torch.Tensor, torch.Tensor] = (bypass_inds, selected_inds)
+    return merged_vecs, merged_clus_labels
+
+
+@torch.jit.script
+def get_closest_embeddings(
+    label_aff_mat: torch.Tensor, target_emb_index: torch.Tensor, merge_quantity: int
+) -> Tuple[torch.Tensor, torch.Tensor]:
+    """
+    Get the indeces of the embedding vectors we want to merge.
```
Comment (SeanNaren) [inline nemo/collections/asr/parts/utils/online_clustering.py:359]:
```suggestion
    Get the indices of the embedding vectors we want to merge.
```

File: tests/collections/asr/test_diar_utils.py [hunk 8]
```diff
@@ -35,130 +47,289 @@ def check_range_values(target, source):
...
 
-def generate_mock_emb(n_emb_per_spk, perturb_sigma, emb_dim):
-    """Generate a set of artificial embedding vectors from random numbers
+def generate_orthogonal_embs(total_spks, perturb_sigma, emb_dim):
+    """Generate a set of artificial orthogonal embedding vectors from random numbers
     """
-    return torch.rand(1, emb_dim).repeat(n_emb_per_spk, 1) + perturb_sigma * torch.rand(n_emb_per_spk, emb_dim)
+    gaus = torch.randn(emb_dim, emb_dim)
+    _svd = torch.linalg.svd(gaus)
+    orth = _svd[0] @ _svd[2]
```
Comment (nithinraok) [inline tests/collections/asr/test_diar_utils.py:70]: Isn't `@` for matrix multiplication removed from python with 3.9?

File: nemo/collections/asr/parts/utils/online_clustering.py [hunk 9]
```diff
@@ -0,0 +1,1106 @@
...
+            For the short audio recordings, clustering algorithm cannot
+            accumulate enough amount of speaker profile for each cluster.
+            Thus, function `getEnhancedSpeakerCount` employs anchor embeddings
+            (dummy representations) to mitigate the effect of cluster sparsity.
+            enhanced_count_thres = 40 is recommended.
+        sparse_search_volume (int):
+            Number of p_values we search during NME analysis.
+            Default is 30. The lower the value, the faster NME-analysis becomes.
+            Lower than 20 might cause a poor parameter estimation.
+        fixed_thres (float):
```
Comment (nithinraok) [inline nemo/collections/asr/parts/utils/online_clustering.py:522]: which threshold is this variable about?

File: nemo/collections/asr/parts/utils/online_clustering.py [hunk 10]
```diff
@@ -0,0 +1,1106 @@
...
+        sparse_search (bool):
+            Toggle sparse search mode. If True, limit the size of p_value_list to sparse_search_volume.
+        cuda (bool):
+            Use cuda for Eigen decomposition if cuda=True.
+        device (torch.device):
+            Torch device variable
+
+    Online Processing Attributes:
+
+        history_buffer_size (int):
```
Comment (nithinraok) [inline nemo/collections/asr/parts/utils/online_clustering.py:539]: Did the default value of 150 come based on experiments results of accuracy or performance?

File: nemo/collections/asr/parts/utils/online_clustering.py [hunk 11]
```diff
@@ -0,0 +1,1106 @@
...
+            - This is a buffer where diarization history is saved in the form of averaged speaker embedding vector.
```
Comment (nithinraok) [inline nemo/collections/asr/parts/utils/online_clustering.py:540]: Can you add in the readme that this value is talking about number of embeddings

File: nemo/collections/asr/parts/utils/online_clustering.py [hunk 12]
```diff
@@ -0,0 +1,1106 @@
...
+        """
+        To save the running time, the p-value is only estimated in the beginning of the session.
+        After switching to online mode, the system uses the most common estimated p-value.
+        Estimating p-value requires a plenty of computational resource. The less frequent estimation of
+        p-value can speed up the clustering algorithm by a huge margin.
+
+        Args:
+            nmesc: (NMESC)
+                nmesc instance.
+            is_online: (bool)
```
Comment (nithinraok) [inline nemo/collections/asr/parts/utils/online_clustering.py:666]: I don;t see this in the argument and frame_index definition is missing

File: nemo/collections/asr/parts/utils/online_clustering.py [hunk 13]
```diff
@@ -0,0 +1,1106 @@
...
+    def __init__(
+        self,
+        max_num_speakers: int,
+        max_rp_threshold: float = 0.15,
+        enhanced_count_thres: float = 40,
+        fixed_thres: float = -1.0,
+        sparse_search_volume: int = 15,
+        history_buffer_size: int = 150,
+        current_buffer_size: int = 150,
+        min_spk_counting_buffer_size=3,
```
Comment (nithinraok) [inline nemo/collections/asr/parts/utils/online_clustering.py:574]: isn't this default value small?

File: nemo/collections/asr/parts/utils/offline_clustering.py [hunk 14]
```diff
@@ -418,12 +424,69 @@ def getCosAffinityMatrix(emb: torch.Tensor) -> torch.Tensor:
...
+    in the range of [-half_scale, half_scale]. The distance to the center of the base-scale is used for
+    calculating the interpolation weight. The distance is squared then normalized to create an interpolation
+    weight vector interpol_w. Using the interpolation weight interpol_w and target embedding indexes target_bool,
+    interpolated embedding is calculated.
+
+    Args:
+        multiscale_weights (Tensor):
+            Tensor containing Multiscale weights
+            Dimensions: (Number of scales) x 1
+        embeddings_in_scales (list):
```
[the remark this answers is not in the document]
Comment (tango4j) [reply nemo/collections/asr/parts/utils/offline_clustering.py:457]: No, base-scale is always the highest scale

File: nemo/collections/asr/parts/utils/online_clustering.py [hunk 15]
```diff
@@ -0,0 +1,1106 @@
...
+        """
+        self.is_online = False
+        self.history_embedding_buffer_emb = torch.tensor([])
+        self.history_embedding_buffer_label = torch.tensor([])
+        self.Y_fullhist = torch.tensor([])
+
+    def onlineNMEanalysis(self, nmesc: NMESC, frame_index: int) -> Tuple[torch.LongTensor, torch.Tensor]:
+        """
+        To save the running time, the p-value is only estimated in the beginning of the session.
+        After switching to online mode, the system uses the most common estimated p-value.
```
[the remark this answers is not in the document]
Comment (tango4j) [reply nemo/collections/asr/parts/utils/online_clustering.py:659]: "The most common" = mode value. ex 1,1,1,1,2,2,3 -> the most common: 1

File: tests/collections/asr/test_diar_utils.py (hunk 8 above)
Comment (tango4j) [reply tests/collections/asr/test_diar_utils.py:70]: As far as I know, `a @ b` is equivalent to `np.matmul(a, b)` and introduced in python 3.5 (quite recently)
Checkout python.org documents: 
https://docs.python.org/3/whatsnew/3.5.html#whatsnew-pep-465

It works with no problem with python 3.9,
but correct me if I am wrong, if it is deprecated.
I don't see any document saying that `@` is deprecated.

File: nemo/collections/asr/parts/utils/online_clustering.py (hunk 12 above)
Comment (tango4j) [reply nemo/collections/asr/parts/utils/online_clustering.py:666]: fixed this docstring

File: nemo/collections/asr/parts/utils/online_clustering.py (hunk 11 above)
Comment (tango4j) [reply nemo/collections/asr/parts/utils/online_clustering.py:540]: Will add that description when README is updated

File: nemo/collections/asr/parts/utils/online_clustering.py (hunk 10 above)
Comment (tango4j) [reply nemo/collections/asr/parts/utils/online_clustering.py:539]: This value is tested on AMI dev. Will update later if it is needed.

File: tests/collections/asr/test_diar_utils.py [hunk 16]
```diff
@@ -1,3 +1,4 @@
+# pylint: disable=no-member
```
[the remark this answers is not in the document]
Comment (tango4j) [reply tests/collections/asr/test_diar_utils.py:1]: Didn't notice. Removed.

File: nemo/collections/asr/parts/utils/online_clustering.py [hunk 17]
```diff
@@ -0,0 +1,1106 @@
...
+        min_samples_for_nmesc (int):
+            The minimum number of samples required for NME clustering. This avoids
+            zero p_neighbour_lists. If the input has fewer segments than min_samples,
+            it is directed to the enhanced speaker counting mode.
+        sparse_search (bool):
+            Toggle sparse search mode. If True, limit the size of p_value_list to sparse_search_volume.
+        cuda (bool):
+            Use cuda for Eigen decomposition if cuda=True.
+        device (torch.device):
+            Torch device variable
```
[the remark this answers is not in the document]
Comment (tango4j) [reply nemo/collections/asr/parts/utils/online_clustering.py]: removed both cuda/device to use the device of the input variables.

File: nemo/collections/asr/parts/utils/offline_clustering.py [hunk 18]
```diff
@@ -331,7 +334,10 @@ def getAffinityGraphMat(affinity_mat_raw: torch.Tensor, p_value: int) -> torch.T
     Calculate a binarized graph matrix and
     symmetrize the binarized graph matrix.
     """
-    X = getKneighborsConnections(affinity_mat_raw, p_value)
+    if p_value <= 0:
+        X = affinity_mat_raw
+    else:
+        X = getKneighborsConnections(affinity_mat_raw, p_value)
```
Comment (SeanNaren) [inline nemo/collections/asr/parts/utils/offline_clustering.py:340]:
```suggestion
    X = affinity_mat_raw if p_value <= 0 else getKneighborsConnections(affinity_mat_raw, p_value)
```

File: nemo/collections/asr/parts/utils/offline_clustering.py [hunk 19]
```diff
@@ -602,14 +666,14 @@ def getEnhancedSpeakerCount(
             max_num_speakers=emb.shape[0],
             max_rp_threshold=0.15,
             sparse_search=True,
-            sparse_search_volume=50,
+            sparse_search_volume=10,
```
Comment (SeanNaren) [inline nemo/collections/asr/parts/utils/offline_clustering.py:669]: n00b question: was this just reduced for performance?

File: nemo/collections/asr/parts/utils/online_clustering.py [hunk 20]
```diff
@@ -0,0 +1,1096 @@
...
+        cuda (bool):
+            Use cuda for Eigen decomposition if cuda=True.
+
+    Online Processing Attributes:
+
+        history_buffer_size (int):
+            - This is a buffer where diarization history is saved in the form of averaged speaker embedding vector.
+            - The values in [50, 200] range is recommended while the system requires bigger buffer size for
+              sessions with larger number of speakers.
+        current_buffer_size (int):
```
Comment (virajkarandikar) [inline nemo/collections/asr/parts/utils/online_clustering.py:540]: what is the unit for buffer size parameters ?

File: nemo/collections/asr/parts/utils/offline_clustering.py (hunk 19 above)
Comment (tango4j) [reply nemo/collections/asr/parts/utils/offline_clustering.py:669]: There is a chance that `sparse_search_volume` could deteriorate the performance if it is too low.

File: nemo/collections/asr/parts/utils/online_clustering.py (hunk 20 above)
Comment (tango4j) [reply nemo/collections/asr/parts/utils/online_clustering.py:540]: It is number of embeddings

**Review by SeanNaren — COMMENTED** [review]
Overall PR looks good, I do not understand everything but I don't think that's needed!

It seems the tests cover the majority of cases, so happy to merge. Will wait for @nithinraok to approve.

**Review by SeanNaren — APPROVED** [review]
